### ISSUE TITLE
unreachable code detection for switch cases out of range of string literal types

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#6149]
```typescript
type foo = 'hey' | 'nay';

switch (<foo>undefined) {
    case 'boom': console.log('never the case'); break; // expected unreachable code, actual no problem
}
```

**Comment by RyanCavanaugh** [comment]
This is somewhat surprising -- we should be issuing an error that `boom` isn't assignable to `'hey'|'nay'`. @DanielRosenwasser what's your take?

**Comment by DanielRosenwasser** [comment]
`"boom"` is a `string`, and `"hey" | "nay"` are assignable to `string`s, so you can check against the two.

We could contextually type `case` expressions, but how exactly is this different from `(<foo>undefined) === "boom"`? We'd need a proposal to catch this.

**Comment by aleksey-bykov** [comment]
In this case `"boom"` isn't just a string (expression), but a string **literal** which we can immediately match against known cases, don't we?

**Comment by DanielRosenwasser** [comment]
Check out #6196 which will make this do what you want.

**Comment by aleksey-bykov** [comment]
I will, thanks a lot!

**Comment by aleksey-bykov** [comment]
it works!
```
let baz : 'foo' | 'bar';
switch (baz) {
   case 'baz': console.log('hey!');break;
}
```
gives
```
test.ts(3,9): error TS2322: Type '"baz"' is not assignable to type '"foo" | "bar"'.
  Type '"baz"' is not assignable to type '"bar"'.
```
looking to see it merged in

**Comment by pocesar** [comment]
1.8 isn't complaining about a declared type of the following: 

```ts
// in a node_module with a typings field
export declare type OrderStatus = 'CREATED' | 'WAITING' | 'PAID' | 'NOT_PAID' | 'REVERTED';
```

```ts
                scope['status'] = (item: IModelTransaction) => {
                    switch (item.status) {  // where > item.status: OrderStatus;
                        case 'CRATED': // not complaining and compiling
                            return 'Criado';
                        case 'PAID':
                            return 'Pago';
                        case 'WAITING':
                            return 'Aguardando';
                        case 'REVERTED':
                            return 'Estornado';
                        case 'NOT_PAID':
                            return 'Não pago';
                    }
                };
```

**Comment by DanielRosenwasser** [comment]
Yes, the issue won't be fixed in 1.8

**Comment by RyanCavanaugh** [comment]
@DanielRosenwasser are the labels on this issue correct?

**Comment by ahejlsberg** [comment]
This issue is fixed with #9407.
